design.md

```diff
@@ -83,41 +180,38 @@ If no script is running and no gRPC interaction has occurred for some time perio
-- All hardware access must go through a robot (i.e., no direct access to hardware).
-  - Scripts would not work in a simulator if they could directly access the hardware.
-- We should use most of the hardware layer from the previous version of the kernel.
+- Use the existing JCSG API for now.
 
-#### RPC
+### Installing and Updating
 
-- We should use the Bowler RPC protocol from the previous version of the kernel.
-  - Need to add support for asking the device its capabilities so that the kernel can figure out if it needs to upload a new program.
+#### Published Artifacts
 
-#### Dependency Management
+! RFC: Refine this
 
-- The kernel must be able to download the toolchain for the device specified in the robot config file. This is because of the [usability requirement](#usability).
-  - The location the toolchain is downloaded from must be stored separately from the source code so that we can update it independently of the kernel. This must also include the version range of the kernel that the toolchain is compatible with. This should be implemented as a GitHub repo containing a registry of toolchains.
+- A shadow Jar that is cross-platform for desktop-class Windows, macOS, and Linux.
+- A shadow Jar for each embedded computer we support.
+- An image for each embedded computer we support.
+- deb and rpm packages that install the auto-updater for the kernel.
 
-#### uC Code
+#### Updating
 
-- The Bowler RPC protocol can handle dynamic definition of device resources, but including support for every possible device resource is not possible because uC's have limited space. Therefore, the kernel must be able to generate, compile, and flash C++ programs to the device.
-- The kernel must be able to determine if it needs to upload a new program to the device. It will do this by asking the device for its capabilities, which will include the supported device resources (see [RPC](#rpc)). If a device resource in the robot config file being loaded is not in the supported capabilities, then the kernel must generate, compile, and flash a new program to the device.
+! RFC: Refine this
 
-### CAD
+- We want to use an auto-updater because there is no cross-platform auto-updating unless we write it ourselves.
+  - The IntelliJ plugin should not auto-update. We should use the plugin marketplace for that.
+  - The kernel and the display can auto-update. If the IntelliJ plugin is installed, the auto-updater shouldn't install a version of the kernel that is incompatible with the plugin. The auto-updater also shouldn't install a version of the display that is incompatible with the kernel.
```

Comment:
gRPC is designed to make backwards-compatibility as automatic as possible, so this shouldn't be too much of an issue.